File: java/leetcode/ShortestPathKeys.java
//Leetcode problem 865 Shortest Path to Get All Keys
//Solution written by Xuqiang Fang on 9 July, 2018
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;
import java.util.PriorityQueue;
import java.util.Stack;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;
class Solution {
    class State {
        int keys, i, j;
        State(int keys, int i, int j) {
            this.keys = keys;
            this.i = i;
            this.j = j;
        }
        @Override
        public int hashCode(){
            return (keys + " " + i + " " + j).hashCode();
        }
    }
    public int shortestPathAllKeys(String[] grid) {
        int x = -1, y = -1, m = grid.length, n = grid[0].length(), max = -1;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                char c = grid[i].charAt(j);
                if (c == '@') {
                    x = i;
                    y = j;
                }
                if (c >= 'a' && c <= 'f') {
                    max = Math.max(c - 'a' + 1, max);
                }
            }
        }
        State start = new State(0, x, y);
        Queue<State> q = new LinkedList<>();
        Set<String> visited = new HashSet<>();
        visited.add(0 + " " + x + " " + y);
        q.offer(start);
        int[][] dirs = new int[][]{{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int step = 0;
        while (!q.isEmpty()) {
            int size = q.size();
            while (size-- > 0) {
                State cur = q.poll();
                System.out.println("step " + step + " (" + cur.i + " , " + cur.j + ")" + cur.keys);
                if (cur.keys == (1 << max) - 1) {
                    return step;
                }
                for (int[] dir : dirs) {
                    int i = cur.i + dir[0];
                    int j = cur.j + dir[1];
                    int keys = cur.keys;
                    if (i >= 0 && i < m && j >= 0 && j < n) {
                        char c = grid[i].charAt(j);
                        if (c == '#') {
                            continue;
                        }
                        if (c >= 'a' && c <= 'f') {
                            keys |= 1 << (c - 'a');
                        }
                        if (c >= 'A' && c <= 'F' && ((keys >> (c - 'A')) & 1) == 0) {
                            continue;
                        }
                        if (!visited.contains(keys + " " + i + " " + j)) {
                            visited.add(keys + " " + i + " " + j);
                            q.offer(new State(keys, i, j));
                        }
                    }
                }
            }
            step++;
        }
        return -1;
    }
}
public class ShortestPathKeys{
	public static void main(String[] args){
		Solution s = new Solution();
        String[] grid = {"@...a",".###A","b.BCc"};
        System.out.println(s.shortestPathAllKeys(grid));
        grid = new String[]{".#.b.","A.#aB","#d...","@.cC.","D...#"};
        System.out.println(s.shortestPathAllKeys(grid));
        grid = new String[]{"@abcdeABCDEFf"};
        System.out.println(s.shortestPathAllKeys(grid));
	}
}

File: py/leetcode/ShortestPathKeys.py
'''
Leetcode problem No 865 Shortest Path to Get All Keys
Solution written by Xuqiang Fang on 8 July, 2018
'''
import collections
class Solution(object):
    def shortestPathAllKeys(self, grid):
        """
        :type grid: List[str]
        :rtype: int
        """
        m = len(grid)
        n = len(grid[0])
        start = (0,0,0); keys = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '@':
                    start = (i, j, 0)
                elif ord(grid[i][j]) >= ord('a') \
                and ord(grid[i][j]) <= ord('f'):
                    keys += 1
        ans = (1 << keys) - 1
        q = collections.deque([start])
        dirs = [(0, 1), (1, 0), (0, -1), (-1,0)]
        visited = set()
        visited.add(start)
        step = 0
        while q:
            t = len(q)
            while t > 0:
                t -= 1
                curr = q.popleft()
                k = curr[2]
                if k == ans:
                    return step
                #print(step , curr)
                for d in dirs:
                    x = curr[0] + d[0]
                    y = curr[1] + d[1]
                    k = curr[2] # watch out that the key may change after this
                    if x>=0 and x<m and y>=0 and y<n:
                        if grid[x][y] == '#':
                            continue
                        value = ord(grid[x][y])
                        if value >= ord('a') and value <= ord('f'):
                            k |= (1 << (value-ord('a'))) # key change here
                        if value >= ord('A') and value <= ord('F') and (k >> (value-ord('A'))&1) == 0:
                            continue
                        if (x,y,k) not in visited:
                            visited.add((x,y,k))
                            q.append((x,y,k))
            step += 1
        return -1
    

def main():
    s = Solution()
    grid = ["@.a.#","###.#","b.A.B"]
    print(s.shortestPathAllKeys(grid))
    grid = ["@..aA","..B#.","....b"]
    print(s.shortestPathAllKeys(grid))
    grid = [".@aA"]
    print(s.shortestPathAllKeys(grid))
    grid = ["@...a",".###A","b.BCc"] # 10
    print(s.shortestPathAllKeys(grid))
    grid = [".#.b.","A.#aB","#d...","@.cC.","D...#"]
    for g in grid:
        print(g)
    print(s.shortestPathAllKeys(grid))
    grid = ["@abcdeABCDEFf"]
    print(s.shortestPathAllKeys(grid))
main()
